polar-core/tests/integration_tests.rs

```diff
@@ -249,34 +264,78 @@ fn qvars(polar: &mut Polar, query_str: &str, vars: &[&str]) -> Vec<Vec<Value>> {
         .collect()
 }
 
-/// Adapted from <http://web.cse.ohio-state.edu/~stiff.4/cse3521/prolog-resolution.html>
-#[test]
-fn test_functions() {
-    let mut polar = Polar::new();
-    polar
-        .load_str("f(1); f(2); g(1); g(2); h(2); k(x) if f(x) and h(x) and g(x);")
-        .unwrap();
+#[track_caller]
+fn qvars(p: &mut Polar, query_str: &str, variables: &[&str], expected: Vec<Vec<Value>>) {
+    assert_eq!(vars(p, query_str, variables), expected);
+}
+
+fn _qruntime(p: &mut Polar, query_str: &str) -> ErrorKind {
+    p.new_query(query_str, false)
+        .unwrap()
+        .next_event()
+        .unwrap_err()
+        .kind
+}
 
-    assert!(qnull(&mut polar, "k(1)"));
-    assert!(qeval(&mut polar, "k(2)"));
-    assert!(qnull(&mut polar, "k(3)"));
-    assert_eq!(qvar(&mut polar, "k(a)", "a"), vec![value!(2)]);
+macro_rules! qruntime {
+    ($query:tt, $err:pat) => {
+        assert!(matches!(_qruntime(&mut Polar::new(), $query), ErrorKind::Runtime($err)));
+    };
+
+    ($query:tt, $err:pat, $cond:expr) => {
```

Comment:
TIOLI: You could make these conditional tokens, something like:

```suggestion
    ($query:tt, $err:pat, $($cond:expr)?) => {
```